dao/src/main/java/org/thingsboard/server/dao/sql/alarm/AlarmRepository.java

```diff
@@ -345,4 +345,7 @@ String updateAlarm(@Param("t_id") UUID tenantId, @Param("a_id") UUID alarmId, @P
 
     @Query(value = "SELECT unassign_alarm(:t_id, :a_id, :a_ts)", nativeQuery = true)
     String unassignAlarm(@Param("t_id") UUID tenantId, @Param("a_id") UUID alarmId, @Param("a_ts") long unassignTime);
+
+    @Query(value = "SELECT DISTINCT a.type FROM alarm a WHERE a.tenant_id = :tenantId LIMIT 256", nativeQuery = true)
```

Comment:
I thought the limit would be the parameter of the requests. Sorry if I did not make myself clear.

Reply:
@smatvienko-tb , please provide your 2 cents on this query. I don't want to do a premature optimization but also don't want to introduce slow queries. The request is cached.

Reply:
@ashvayka `SELECT DISTINCT` is a slow query as it acts as `GROUP BY` query. Limit 256 is about to limit result query (256 uniq types), but not a sample 256 rows to find unique types.

Reply:
the fastest way to fetch such data is by having a separate table 
alarm_type by tenant
and full-text index by trigrams
https://www.postgresql.org/docs/current/pgtrgm.html
Example:

CREATE TABLE test_trgm (t text);
CREATE INDEX trgm_idx ON test_trgm USING GIST (t gist_trgm_ops);

or

CREATE INDEX trgm_idx ON test_trgm USING GIN (t gin_trgm_ops);

Reply:
how do we populate and how do we remove records from the alarm_type table in such case?

Reply:
Upsert with returning alarm_type_id first query
insert alarm

Removing is expensive. I guess we can left it with no delete